packages/transaction-controller/src/TransactionController.ts

```diff
@@ -2303,6 +2310,186 @@ export class TransactionController extends BaseController<
+      this.#onTransactionParamsUpdated(
+        transactionMeta,
+        updatedTransactionParams,
+      );
+    }
+
+    return transactionMeta;
+  }
+
+  #checkIfTransactionParamsUpdated(newTransactionMeta: TransactionMeta) {
+    const { id: transactionId, txParams: newParams } = newTransactionMeta;
+
+    const originalParams = this.getTransaction(transactionId)?.txParams;
+
+    if (!originalParams || isEqual(originalParams, newParams)) {
+      return [];
+    }
+
+    const params = Object.keys(newParams) as (keyof TransactionParams)[];
+
+    const updatedProperties = params.filter(
+      (param) => newParams[param] !== originalParams[param],
+    );
+
+    log(
+      'Transaction parameters have been updated',
+      transactionId,
+      updatedProperties,
+      originalParams,
+      newParams,
+    );
+
+    return updatedProperties;
+  }
+
+  #onTransactionParamsUpdated(
+    transactionMeta: TransactionMeta,
+    updatedParams: (keyof TransactionParams)[],
+  ) {
+    if (
+      (['to', 'value', 'data'] as const).some((param) =>
+        updatedParams.includes(param),
+      )
+    ) {
+      log('Updating simulation data due to transaction parameter update');
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      this.#updateSimulationData(transactionMeta);
+    }
+  }
+
+  async #updateSimulationData(transactionMeta: TransactionMeta) {
+    const { id: transactionId, chainId, txParams } = transactionMeta;
+    const { from, to, value, data } = txParams;
+
+    let simulationData: SimulationData = {
+      error: {
+        code: SimulationErrorCode.Disabled,
+        message: 'Simulation disabled',
+      },
+      tokenBalanceChanges: [],
+    };
+
+    if (this.#isSimulationEnabled()) {
+      this.#updateTransactionInternal(
+        { transactionId, skipHistory: true },
+        (txMeta) => {
+          txMeta.simulationData = undefined;
+        },
+      );
+
+      simulationData = await getSimulationData({
+        chainId,
+        from: from as Hex,
+        to: to as Hex,
+        value: value as Hex,
+        data: data as Hex,
+      });
+    }
+
+    const finalTransactionMeta = this.getTransaction(transactionId);
+
+    if (!finalTransactionMeta) {
+      log(
+        'Cannot update simulation data as transaction not found',
+        transactionId,
+        simulationData,
+      );
+
+      return;
+    }
+
+    this.#updateTransactionInternal(
+      {
+        transactionId,
+        note: 'TransactionController#updateSimulationData - Update simulation data',
+      },
+      (txMeta) => {
+        txMeta.simulationData = simulationData;
+      },
+    );
+
+    log('Updated simulation data', transactionId, simulationData);
+  }
+
+  #updateWithCallback(callback: (state: TransactionState) => TransactionState) {
+    const currentState = cloneDeep(this.state);
+    // eslint-disable-next-line n/callback-return
+    const updatedState = callback(currentState);
```

Comment:
Sorry if I miss something here but removing the `return` from inside callback in `updateTransactionInternal` evaluates `void ?? currentState` which will throw an error since `void` is not a value.